--- beecrypt-4.2.1/acinclude.m4.bero	2012-11-25 20:14:19.883818617 +0100
+++ beecrypt-4.2.1/acinclude.m4	2012-11-25 22:40:43.801004540 +0100
@@ -128,6 +128,9 @@ AC_DEFUN([BEE_WITH_ARCH],[
       ;;
     esac
     ;;
+  arm*)
+    bc_target_arch=arm
+    ;;
   esac
   ])
 
@@ -1408,7 +1411,7 @@ AC_DEFUN([BEE_ASM_SOURCES],[
   echo > sha1opt.s
   if test "$ac_enable_debug" != yes; then
     case $bc_target_arch in
-    arm)
+    arm*)
       AC_CONFIG_COMMANDS([mpopt.arm],[
         m4 $srcdir/gas/mpopt.arm.m4 > mpopt.s
         ])
@@ -1610,7 +1610,7 @@ AC_DEFUN([BEE_THREAD_LOCAL_STORAGE],[
   AH_TEMPLATE([ENABLE_THREAD_LOCAL_STORAGE],[Define to 1 if you want to enable thread-local-storage support])
   if test "$ac_enable_threads" = yes; then
     AC_MSG_CHECKING([if your compiler supports thread-local-storage])
-    AC_COMPILE_IFELSE([__thread int a = 0;],[
+    AC_COMPILE_IFELSE([AC_LANG_SOURCE([__thread int a = 0;])],[
       AC_DEFINE([ENABLE_THREAD_LOCAL_STORAGE],1)
       AC_MSG_RESULT([yes])
       ],[
--- beecrypt-4.2.1/configure.ac.bero	2012-11-25 20:14:00.902055919 +0100
+++ beecrypt-4.2.1/configure.ac	2012-11-25 22:37:33.192387475 +0100
@@ -482,6 +482,11 @@ cygwin* | mingw*)
   AC_MSG_CHECKING([for console])
   AC_MSG_RESULT(yes)
   ;;
+linux-android*)
+  # Android doesn't provide sys/soundcard.h
+  # and in general doesn't have OSS or OSS
+  # emulation enabled - so we shouldn't set HAVE_DEV_DSP
+  ;;
 linux*)
   AC_CACHE_CHECK([for /dev/dsp],ac_cv_have_dev_dsp,[
     if test -r /dev/dsp; then
