--- gcc/libcpp/charset.c.bero	2013-04-19 11:15:06.103368432 +0200
+++ gcc/libcpp/charset.c	2013-04-19 11:15:08.079343729 +0200
@@ -1696,7 +1696,7 @@ _cpp_interpret_identifier (cpp_reader *p
 uchar * 
 _cpp_convert_input (cpp_reader *pfile, const char *input_charset,
 		    uchar *input, size_t size, size_t len,
-		    const unsigned char **buffer_start, off_t *st_size)
+		    const unsigned char **buffer_start, loff_t *st_size)
 {
   struct cset_converter input_cset;
   struct _cpp_strbuf to;
--- gcc/libcpp/internal.h.bero	2013-04-19 11:15:15.248254105 +0200
+++ gcc/libcpp/internal.h	2013-04-19 11:15:19.296203499 +0200
@@ -732,7 +732,7 @@ extern cppchar_t _cpp_valid_ucn (cpp_rea
 extern void _cpp_destroy_iconv (cpp_reader *);
 extern unsigned char *_cpp_convert_input (cpp_reader *, const char *,
 					  unsigned char *, size_t, size_t,
-					  const unsigned char **, off_t *);
+					  const unsigned char **, loff_t *);
 extern const char *_cpp_default_encoding (void);
 extern cpp_hashnode * _cpp_interpret_identifier (cpp_reader *pfile,
 						 const unsigned char *id,
--- gcc/libcpp/macro.c.bero	2013-04-19 11:11:28.386090272 +0200
+++ gcc/libcpp/macro.c	2013-04-19 11:12:10.279566532 +0200
@@ -246,7 +246,7 @@ _cpp_builtin_macro_text (cpp_reader *pfi
 		struct tm *tb = NULL;
 		struct stat *st = _cpp_get_file_stat (file);
 		if (st)
-		  tb = localtime (&st->st_mtime);
+		  tb = localtime ((const time_t*)&st->st_mtime);
 		if (tb)
 		  {
 		    char *str = asctime (tb);
