--- gcc/gcc/system.h.bero	2012-10-10 16:42:07.513098286 +0200
+++ gcc/gcc/system.h	2012-10-10 16:43:36.399103334 +0200
@@ -99,7 +99,7 @@ along with GCC; see the file COPYING3.
 extern "C" {
 #endif
 
-# ifdef HAVE_CLEARERR_UNLOCKED
+# if 0 //def HAVE_CLEARERR_UNLOCKED
 #  undef clearerr
 #  define clearerr(Stream) clearerr_unlocked (Stream)
 #  if defined (HAVE_DECL_CLEARERR_UNLOCKED) && !HAVE_DECL_CLEARERR_UNLOCKED
@@ -148,14 +148,14 @@ extern char *fgets_unlocked (char *, int
 extern int fputs_unlocked (const char *, FILE *);
 #  endif
 # endif
-# ifdef HAVE_FERROR_UNLOCKED
+# if 0 //def HAVE_FERROR_UNLOCKED
 #  undef ferror
 #  define ferror(Stream) ferror_unlocked (Stream)
 #  if defined (HAVE_DECL_FERROR_UNLOCKED) && !HAVE_DECL_FERROR_UNLOCKED
 extern int ferror_unlocked (FILE *);
 #  endif
 # endif
-# ifdef HAVE_FREAD_UNLOCKED
+# if 0 //def HAVE_FREAD_UNLOCKED
 #  undef fread
 #  define fread(Ptr, Size, N, Stream) fread_unlocked (Ptr, Size, N, Stream)
 #  if defined (HAVE_DECL_FREAD_UNLOCKED) && !HAVE_DECL_FREAD_UNLOCKED
--- gcc/libgcc/unwind-arm-common.inc.bero	2012-10-10 18:58:02.487562312 +0200
+++ gcc/libgcc/unwind-arm-common.inc	2012-10-10 18:58:11.877562853 +0200
@@ -26,6 +26,7 @@
 #include "unwind.h"
 
 /* Used for SystemTap unwinder probe.  */
+#undef HAVE_SYS_SDT_H
 #ifdef HAVE_SYS_SDT_H
 #include <sys/sdt.h>
 #endif
--- gcc/libiberty/getpagesize.c.bero	2012-10-10 16:29:47.765056194 +0200
+++ gcc/libiberty/getpagesize.c	2012-10-10 16:29:57.701056760 +0200
@@ -60,11 +60,13 @@ BUGS
 # endif /* PAGESIZE */
 #endif /* GNU_OUR_PAGESIZE */
 
+#ifndef __BIONIC__
 int
 getpagesize (void)
 {
   return (GNU_OUR_PAGESIZE);
 }
+#endif
 
 #else /* VMS */
 
