--- vim73/src/auto/configure.bero	2012-10-11 00:31:21.409700268 +0200
+++ vim73/src/auto/configure	2012-10-11 11:11:56.806333958 +0200
@@ -9823,7 +9823,7 @@ fi
 
     if test "x$olibs" != "x$LIBS"; then
                         if test "$cross_compiling" = yes; then :
-  res="FAIL"
+  res="OK"
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
@@ -11642,7 +11642,7 @@ _ACEOF
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking uint32_t is 32 bits" >&5
 $as_echo_n "checking uint32_t is 32 bits... " >&6; }
 if test "$cross_compiling" = yes; then :
-  as_fn_error "could not compile program using uint32_t." "$LINENO" 5
+  $as_echo "$as_me:${as_lineno-$LINENO}: crosscompiling -- assuming ok" >&6
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
